File: 885.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        vector<vector<int>> ans;
        ans.reserve(rows * cols); // Reserve space in advance
        ans.push_back({rStart, cStart});
        
        int radius = 1, allBlocks = rows * cols;
        while(ans.size() < allBlocks) {
            // go right
            for(int i = 0; i < radius; i++) {
                cStart++;
                if (cStart >= 0 && cStart < cols && rStart >= 0 && rStart < rows)
                    ans.push_back({rStart, cStart});
            }
            // go down
            for(int i = 0; i < radius; i++) {
                rStart++;
                if (cStart >= 0 && cStart < cols && rStart >= 0 && rStart < rows)
                    ans.push_back({rStart, cStart});
            }
            radius++;
            // go left
            for(int i = 0; i < radius; i++) {
                cStart--;
                if (cStart >= 0 && cStart < cols && rStart >= 0 && rStart < rows)
                    ans.push_back({rStart, cStart});
            }
            // go up
            for(int i = 0; i < radius; i++) {
                rStart--;
                if (cStart >= 0 && cStart < cols && rStart >= 0 && rStart < rows)
                    ans.push_back({rStart, cStart});
            }
            radius++;
        }
        return ans;
    }
};

int main(){
    Solution s;
    int rows, cols, rStart, cStart;
    cin >> rows >> cols >> rStart >> cStart;
    vector<vector<int>> ans = s.spiralMatrixIII(rows, cols, rStart, cStart);
    for(auto i : ans) {
        cout << i[0] << " " << i[1] << endl;
    }
    return 0;
}
   

File: python/885.py
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> list[list[int]]:
        ans = [[rStart, cStart]]
        all_blocks = rows * cols
        radius = 1
        
        while len(ans) < all_blocks:
            # go right
            for _ in range(radius):
                cStart += 1
                if 0 <= rStart < rows and 0 <= cStart < cols:
                    ans.append([rStart, cStart])
            # go down
            for _ in range(radius):
                rStart += 1
                if 0 <= rStart < rows and 0 <= cStart < cols:
                    ans.append([rStart, cStart])
            radius += 1
            # go left
            for _ in range(radius):
                cStart -= 1
                if 0 <= rStart < rows and 0 <= cStart < cols:
                    ans.append([rStart, cStart])
            # go up
            for _ in range(radius):
                rStart -= 1
                if 0 <= rStart < rows and 0 <= cStart < cols:
                    ans.append([rStart, cStart])
            radius += 1
        
        return ans

if __name__ == '__main__':
    rows, cols, rStart, cStart = map(int, input().split()) 
    print(Solution().spiralMatrixIII(rows, cols, rStart, cStart))